### ISSUE TITLE
get-command-info for system-installed git-subrepo

### ISSUE THREAD

**Issue by spog** [ingydotnet/git-subrepo#472]
Hi!

Would it be possible to replace "readlink" commands for "$lib/git-subrepo.d/upstream" and "$lib/git-subrepo.d/commit" with "cat", for example. Namely it is hard to create an file named "https://github.com/ingydotnet/git-subrepo.git", that would have been poinetd by the "upstream" link. 
Is there any specific reason to use "readlink" command to get info about installed "git-subrepo" command?

regards, Samo

**Comment by admorgan** [comment]
Is there a specific issue you are having with readlink? More importantly readlink just finds the the full path to a file, it is not looking for a file formatted as you suggest. the upstream and commit files are created during make install. If you are using a git install then it will use git to extract the information.

Coming back around what is your symptom?

**Comment by spog** [comment]
I am sorry, i skipped the fact, that readlink result get executed afterwards. Should i just close this issue?

**Comment by admorgan** [comment]
You are correct. So you are worried about the security implications? I agree with this concern.

**Comment by admorgan** [comment]
I am still a little confused, are you not seeing the correct result in the commit message?

**Comment by spog** [comment]
Using my DEB installed "git-subrepo" the ... subrepo pull command writes:
    git-subrepo:
...
      version:  "0.4.1"
      origin:   "???"
      commit:   "???"

And i want to fix those question marks? Maybe i missed something in "make install"?

**Comment by admorgan** [comment]
That was the symptom I was looking for. I can help look into that.

**Comment by spog** [comment]
Those two links "upstream" and "commit"  are missing in "/usr/lib/git-core/git-subrepo.d". As i cheched the code, i thought, i'd generate them from my DEB packaging project, as i did not found any existing generation of them.

**Comment by spog** [comment]
I also thought that generation of this info should to be provided by the packaging code.

**Comment by admorgan** [comment]
As this has never been packaged, it was required to be retrieved from the method of install on the website. I agree that the packager should include this information explicitly.

**Comment by spog** [comment]
Here is what is installed with DEB:
$ dpkg -S git-subrepo
git-subrepo: /usr/lib/git-core/git-subrepo.d/bash+.bash
git-subrepo: /usr/share/man/man1/git-subrepo.1.gz
git-subrepo: /usr/share/bash-completion/completions/git-subrepo
git-subrepo: /usr/lib/git-core/git-subrepo
git-subrepo: /usr/share/doc/git-subrepo
git-subrepo: /usr/share/doc/git-subrepo/changelog.gz
git-subrepo: /usr/lib/git-core/git-subrepo.d
git-subrepo: /usr/share/doc/git-subrepo/copyright
git-subrepo: /usr/lib/git-core/git-subrepo.d/help-functions.bash
$

**Comment by admorgan** [comment]
I see the issue, there should be a /usr/lib/git-core/git-subrepo.d/upstream and /usr/lib/git-core/git-subrepo.d/commit included in the package.

**Comment by spog** [comment]
I can generate two scripts in the packaging code from internal "git-subrepo" "subrepo" (from the .gitrepo file pointing to original code) and add links "upstream" and "commit" to them.

**Comment by spog** [comment]
There might still be a little problem. The executed result from "readline" need to be in a PATH or needs to give absolute path in result (readline -f, if those two were regular files). Symlinks could however jield absolute path, but packaging constraints report problems 
(i.e. W: git-subrepo: symlink-should-be-relative usr/lib/git-core/git-subrepo.d/upstream /usr/lib/git-core/git-subrepo.d/cmd_remote).

Creating regular files (executable scripts) might work just fine, if "readline -f" were used?

**Comment by spog** [comment]
By generating these two simple files:
$ cat /usr/lib/git-core/git-subrepo.d/upstream 
https://github.com/ingydotnet/git-subrepo.git
$ 
$ cat /usr/lib/git-core/git-subrepo.d/commit 
a04d8c2
$ 

and replacing "readlink" with "cat", commits look great:
i.e.
commit 77c0e150492c81e0fba96434ab64159d4ab45a29 (HEAD -> master)
Author: Samo Pogačnik <samo_pogacnik@t-2.net>
Date:   Fri May 29 19:07:20 2020 +0200

    git subrepo clone https://github.com/spog/u2up-cli src4
    
    subrepo:
      subdir:   "src4"
      merged:   "0bedc81"
    upstream:
      origin:   "https://github.com/spog/u2up-cli"
      branch:   "master"
      commit:   "0bedc81"
    git-subrepo:
      version:  "0.4.1"
      origin:   "https://github.com/ingydotnet/git-subrepo.git"
      commit:   "a04d8c2"

-----------
diff --git a/lib/git-subrepo b/lib/git-subrepo
index 3865570..fa248f9 100755
--- a/lib/git-subrepo
+++ b/lib/git-subrepo
@@ -1622,8 +1622,8 @@ get-command-info() {
     # If `git-subrepo` was system-installed (`make install`):
     if [[ -e $lib/git-subrepo.d/upstream ]] &&
        [[ -e $lib/git-subrepo.d/commit ]]; then
-      command_remote=$(readlink "$lib/git-subrepo.d/upstream")
-      command_commit=$(readlink "$lib/git-subrepo.d/commit")
+      command_remote=$(cat "$lib/git-subrepo.d/upstream")
+      command_commit=$(cat "$lib/git-subrepo.d/commit")
     elif [[ $lib =~ / ]]; then
       lib="$(dirname "$lib")"
       if [[ -d $lib/.git ]]; then

**Comment by admorgan** [comment]
I am finally getting back to this and there are competing requirements, one for packagers and one for installers from git and one for extracting the source from a zip.

The current Makefile doesn't create the upstream or commit links stated in those lines and are effectively non-functional. That gives us a great deal of options to work around this issue. The hard part is what data to include. The install via git is easiest because contains the SHA and remote and will remain the same. We already include the version, so we could choose to exclude the source info for the zip and install situation. For packaging would it be better to act like the unzip and install or would you want to include files that expressed your internal source info?

**Comment by anjos** [comment]
From a packager's perspective, removing repo and commit sha from the commit message would make the packaging simpler.  Currently, I have to execute [this little hack](https://github.com/conda-forge/git-subrepo-feedstock/blob/main/recipe/recipe.yaml) to get those fields:

```bash
remote="https://github.com/ingydotnet/git-subrepo"
version="0.4.9"
echo "$(git ls-remote ${remote} --tags ${version} | sed 's/^\([0-9a-f]\{7\}\).*/\1/')" > $PREFIX/libexec/git-core/git-subrepo.d/commit
```

Beyond, of course, the [`readlink` -> `cat` patch](https://github.com/conda-forge/git-subrepo-feedstock/blob/main/recipe/get-info-with-cat.patch) mentioned above.  A second option would be to execute such a script during tarball/zip creation such that the values are always there when the project is shipped.

**Comment by admorgan** [comment]
That is a good idea about creating the files during compression. Do any packagers pull via git instead of the compressed files?

Creating the files during compression only helps, so I will do that and see if there is additional need later.

**Comment by spog** [comment]
I prefer generating orig.tar.gz for debian packaging out of repo tags (preferably signed) instead of using github's release assets.

**Comment by admorgan** [comment]
The information is purely informational, and not really beneficial unless you are using a version that isn't tagged. @spog what is your opinion of not including it unless you are using the git version?

**Comment by spog** [comment]
I think that inclusion of git-subrepo info is primarily beneficial for archive users, as they can not retrieve it like the git users do. On the other side it is not much harm done if missing in archives, i believe. As you mentioned, this info is purely informational and perhaps packagers should provide their own info anyway?

**Comment by anjos** [comment]
My understanding is: It is fine to not have git-subrepo info in the archive, and remove those fields from the commit message in case that is true.  If the commit message must contain those fields, then the information needs to be packed with the archive.

**Comment by spog** [comment]
Each distro policy may require specific modifications of the the original code, which automatically invalidates at least the commit id provided by the original archive.
